src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/ExecutionGraph.java

```diff
@@ -0,0 +1,137 @@
+/* Copyright 2021 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.floodlight.command.rulemanager;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.Data;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ExecutionGraph {
+    private Map<String, Node> nodes = new HashMap<>();
+    private boolean ready;
+    private List<String> topologicalOrder = new ArrayList<>();
+    @VisibleForTesting
+    List<List<String>> stages = new ArrayList<>();
+    private int currentStage = 0;
+
+
+    /**
+     * Gets current stage.
+     */
+    public List<String> getCurrent() {
+        if (!ready) {
+            buildStages();
+        }
+        return stages.get(currentStage);
+    }
+
+    public boolean hasNext() {
+        return currentStage + 1 < stages.size();
+    }
+
+    public void next() {
```

Comment:
Now `getCurrent()` can return null. If `hasNext` and `next` will be merged, i.e. next will switch only if next stage exist. We can get rid of null in response. It still requires carefful treatment of returned succession flag from updated `next()` call, but it looks better than null response from method where you do not check this null.

Or we can unconditionally call `next()` and check `getCurrent()` result on null. It will eliminate any existing API ambiguity.

I am trying to say there is too many tool to controll state of this object.